plutus-contract/test/Spec/Contract/Tx/Constraints/RequiredSigner.hs

```diff
@@ -156,3 +169,98 @@ mustBeSignedByTypedValidator = Scripts.mkTypedValidator @UnitTest
+
+
+{-# INLINABLE mustReferenceOutputValidatorV2 #-}
+mustReferenceOutputValidatorV2 :: PV2.Validator
+mustReferenceOutputValidatorV2 = PV2.mkValidatorScript
+    $$(PlutusTx.compile [|| wrap ||])
+ where
+     mkMustReferenceOutputV2Validator = mkMustReferenceOutputValidator Cons.V2.checkScriptContext
+     wrap = PSU.V2.mkUntypedValidator mkMustReferenceOutputV2Validator
+
+tag :: Trace.ContractInstanceTag
+tag = "instance 1"
+
+mkMustReferenceOutputValidator
+    :: (Constraints.TxConstraints Void Void -> sc -> Bool)
+    -> PV1.TxOutRef -> () -> sc -> Bool
+mkMustReferenceOutputValidator checkScriptContext txOutRef _ =
+    checkScriptContext (Constraints.mustReferenceOutput txOutRef)
+
+mustReferenceOutputV2ValidatorAddress :: Ledger.Address
+mustReferenceOutputV2ValidatorAddress =
+    PSU.V2.mkValidatorAddress mustReferenceOutputValidatorV2
+
+cardanoTxOwnWalletContract
+    :: Ledger.PaymentPubKey
+    -> Ledger.PaymentPubKeyHash
+    -> Contract () EmptySchema ContractError ()
+cardanoTxOwnWalletContract pk pkh = do
+    let mkTx lookups constraints = either (error . show) id $ TxCons.mkTx @Any def lookups constraints
+
+    utxos <- ownUtxos
+    myAddr <- ownAddress
+    let get3 (a:b:c:_) = (a, b, c)
+        get3 _         = error "Spec.Contract.TxConstraints.get3: not enough inputs"
+        ((utxoRef, utxo), (utxoRefForBalance1, _), (utxoRefForBalance2, _)) = get3 $ M.toList utxos
+        vh = fromJust $ Ledger.toValidatorHash mustReferenceOutputV2ValidatorAddress
+        lookups1 = Constraints.unspentOutputs utxos
+               <> Constraints.plutusV2OtherScript mustReferenceOutputValidatorV2
+               <> Constraints.paymentPubKey pk
+        tx1 = Constraints.mustPayToOtherScriptWithDatumInTx
+                vh
+                (Ledger.Datum $ PlutusTx.toBuiltinData utxoRef)
+                (Ada.adaValueOf 5)
+          <> Constraints.mustSpendPubKeyOutput utxoRefForBalance1
+          <> Constraints.mustUseOutputAsCollateral utxoRefForBalance1
+          <> Constraints.mustPayToAddressWithReferenceValidator
+                myAddr
+                vh
+                Nothing
+                (Ada.adaValueOf 30)
+    submitTxConfirmed $ mkTx lookups1 tx1
+
+    -- Trying to unlock the Ada in the script address
+    scriptUtxos <- utxosAt mustReferenceOutputV2ValidatorAddress
+    utxos' <- ownUtxos
+    let
+        scriptUtxo = fst . head . M.toList $ scriptUtxos
+        refScriptUtxo = head . M.keys . M.filter (has $ Tx.decoratedTxOutReferenceScript . _Just) $ utxos'
+        lookups2 = Constraints.unspentOutputs (M.singleton utxoRef utxo <> scriptUtxos <> utxos')
+        tx2 = Constraints.mustReferenceOutput utxoRef
+          <> Constraints.mustSpendScriptOutputWithReference scriptUtxo Ledger.unitRedeemer refScriptUtxo
+          <> Constraints.mustSpendPubKeyOutput utxoRefForBalance2
+          <> Constraints.mustUseOutputAsCollateral utxoRefForBalance2
+          <> Constraints.mustBeSignedBy pkh
```

Comment:
Need to use `mustBeSignedBy` constraint to embed pkh in unblaanced tx body. No need to use reference inputs or explicit collateral input.

Reply:
ignore this, clearly being used.